File: paseto/paseto.go
// Package paseto implements PASETO v4.public — Ed25519-signed public tokens —
// over the standard library only (ADR-0002: format layers are hand-written,
// primitives never are).
//
// v4.public token layout:
//
//	v4.public.<b64url(message || signature)>[.<b64url(footer)>]
//
// where signature = Ed25519-Sign(sk, PAE([h, m, f, i])), h the header
// including trailing dot, m the message, f the footer ("" when absent) and i
// the implicit assertion. PAE is the spec's Pre-Authentication Encoding; it
// makes the signed byte string injective over its pieces, which is what stops
// a footer byte from being reinterpreted as a message byte.
package paseto

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"strings"
)

const headerV4Public = "v4.public."

var (
	ErrMalformed   = errors.New("paseto: malformed token")
	ErrWrongHeader = errors.New("paseto: not a v4.public token")
	ErrBadSig      = errors.New("paseto: signature verification failed")
	ErrKeySize     = errors.New("paseto: wrong key size")
)

var b64 = base64.RawURLEncoding.Strict()

// pae implements Pre-Authentication Encoding from the PASETO spec:
// LE64(count) || for each piece: LE64(len(piece)) || piece.
// LE64 encodes a uint64 little-endian with the most significant bit cleared.
func pae(pieces ...[]byte) []byte {
	size := 8
	for _, p := range pieces {
		size += 8 + len(p)
	}
	out := make([]byte, 0, size)
	out = le64(out, uint64(len(pieces)))
	for _, p := range pieces {
		out = le64(out, uint64(len(p)))
		out = append(out, p...)
	}
	return out
}

func le64(dst []byte, n uint64) []byte {
	var b [8]byte
	binary.LittleEndian.PutUint64(b[:], n&0x7FFFFFFFFFFFFFFF)
	return append(dst, b[:]...)
}

// Sign produces a v4.public token over message with an optional footer and
// implicit assertion. The implicit assertion is authenticated but NOT carried
// in the token; both sides must agree on it out of band.
func Sign(sk ed25519.PrivateKey, message, footer, implicit []byte) (string, error) {
	if len(sk) != ed25519.PrivateKeySize {
		return "", ErrKeySize
	}
	m2 := pae([]byte(headerV4Public), message, footer, implicit)
	sig := ed25519.Sign(sk, m2)

	var b strings.Builder
	b.WriteString(headerV4Public)
	b.WriteString(b64.EncodeToString(append(append([]byte{}, message...), sig...)))
	if len(footer) > 0 {
		b.WriteByte('.')
		b.WriteString(b64.EncodeToString(footer))
	}
	return b.String(), nil
}

// Parse splits a token without verifying it. Callers must treat the result as
// untrusted until Verify succeeds; it exists so the kid can be read from a
// footer to select a verification key.
func Parse(token string) (message, signature, footer []byte, err error) {
	if !strings.HasPrefix(token, headerV4Public) {
		return nil, nil, nil, ErrWrongHeader
	}
	rest := token[len(headerV4Public):]
	var bodyPart, footPart string
	if i := strings.IndexByte(rest, '.'); i >= 0 {
		bodyPart, footPart = rest[:i], rest[i+1:]
		if footPart == "" || strings.IndexByte(footPart, '.') >= 0 {
			return nil, nil, nil, ErrMalformed
		}
	} else {
		bodyPart = rest
	}
	body, err := b64.DecodeString(bodyPart)
	if err != nil || len(body) < ed25519.SignatureSize {
		return nil, nil, nil, ErrMalformed
	}
	if footPart != "" {
		footer, err = b64.DecodeString(footPart)
		if err != nil {
			return nil, nil, nil, ErrMalformed
		}
	}
	n := len(body) - ed25519.SignatureSize
	return body[:n], body[n:], footer, nil
}

// Verify checks the token against pk and returns the message on success.
func Verify(pk ed25519.PublicKey, token string, implicit []byte) (message, footer []byte, err error) {
	if len(pk) != ed25519.PublicKeySize {
		return nil, nil, ErrKeySize
	}
	m, sig, f, err := Parse(token)
	if err != nil {
		return nil, nil, err
	}
	m2 := pae([]byte(headerV4Public), m, f, implicit)
	if !ed25519.Verify(pk, m2, sig) {
		return nil, nil, ErrBadSig
	}
	return m, f, nil
}

File: node/src/paseto.ts
/**
 * PASETO v4.public — Ed25519-signed tokens — over node:crypto alone.
 *
 * Token layout:
 *
 *   v4.public.<b64url(message || signature)>[.<b64url(footer)>]
 *
 * where signature = Ed25519-Sign(sk, PAE([h, m, f, i])). PAE is the spec's
 * Pre-Authentication Encoding; it makes the signed byte string injective over
 * its pieces, which is what stops a footer byte being reinterpreted as a
 * message byte.
 */
import { createPublicKey, verify as cryptoVerify, type KeyObject } from "node:crypto";

const HEADER = "v4.public.";
const SIGNATURE_BYTES = 64;
const PUBLIC_KEY_BYTES = 32;

export class PasetoError extends Error {}

export function b64urlDecode(s: string): Buffer {
  // Strict: a token is generated, never typed, so padding or a stray
  // character means tampering rather than a formatting preference.
  if (!/^[A-Za-z0-9_-]*$/.test(s)) throw new PasetoError("paseto: malformed token");
  return Buffer.from(s, "base64url");
}

export function b64urlEncode(b: Buffer | Uint8Array): string {
  return Buffer.from(b).toString("base64url");
}

/** LE64 with the most significant bit cleared, per the PASETO spec. */
function le64(n: number): Buffer {
  const b = Buffer.alloc(8);
  b.writeBigUInt64LE(BigInt(n) & 0x7fffffffffffffffn);
  return b;
}

export function pae(...pieces: Buffer[]): Buffer {
  const out: Buffer[] = [le64(pieces.length)];
  for (const p of pieces) {
    out.push(le64(p.length), p);
  }
  return Buffer.concat(out);
}

export interface ParsedToken {
  message: Buffer;
  signature: Buffer;
  footer: Buffer;
}

/**
 * Split a token WITHOUT verifying it.
 *
 * The result is untrusted until verify() succeeds. It exists so the kid can be
 * read from the footer to select a verification key — a read that may only
 * ever index a bounded, already-loaded map.
 */
export function parse(token: string): ParsedToken {
  if (!token.startsWith(HEADER)) throw new PasetoError("paseto: not a v4.public token");
  const rest = token.slice(HEADER.length);
  const dot = rest.indexOf(".");
  let bodyPart = rest;
  let footerPart = "";
  if (dot >= 0) {
    bodyPart = rest.slice(0, dot);
    footerPart = rest.slice(dot + 1);
    if (footerPart === "" || footerPart.includes(".")) throw new PasetoError("paseto: malformed token");
  }
  const body = b64urlDecode(bodyPart);
  if (body.length < SIGNATURE_BYTES) throw new PasetoError("paseto: malformed token");
  const footer = footerPart ? b64urlDecode(footerPart) : Buffer.alloc(0);
  const cut = body.length - SIGNATURE_BYTES;
  return { message: body.subarray(0, cut), signature: body.subarray(cut), footer };
}

/** Build a KeyObject from the 32 raw public key bytes the key document carries. */
export function publicKeyFromRaw(raw: Buffer): KeyObject {
  if (raw.length !== PUBLIC_KEY_BYTES) throw new PasetoError("paseto: wrong key size");
  return createPublicKey({
    key: { kty: "OKP", crv: "Ed25519", x: raw.toString("base64url") },
    format: "jwk",
  });
}

/** Verify a token and return its message. Throws on any failure. */
export function verify(key: KeyObject, token: string, implicit: Buffer = Buffer.alloc(0)): ParsedToken {
  const parsed = parse(token);
  const m2 = pae(Buffer.from(HEADER), parsed.message, parsed.footer, implicit);
  // Ed25519 takes a null algorithm: the curve fixes the hash.
  if (!cryptoVerify(null, m2, key, parsed.signature)) {
    throw new PasetoError("paseto: signature verification failed");
  }
  return parsed;
}
